README.rst

```diff
@@ -0,0 +1,329 @@
+
+      createDataset("purchaseRecords", KeyValueTable.class);
+      createDataset("userPurchases", KeyValueTable.class);
+      createDataset("productPurchases", KeyValueTable.class);
+    }
+  }
+  
+The ``PurchaseWorkflowApp`` application defines a new `Stream
+<http://docs.cdap.io/cdap/current/en/developers-manual/building-blocks/streams.html>`__
+where purchase events are ingested. Once the data is
+ingested, the events can be processed in real time or batch. In our
+application, we will process the events in batch using the
+``PurchaseWorkflow`` program and compute the total purchases made by a particular user
+and the total purchases made for a particular product in a specific time range. We will use three MapReduce
+programs ``PurchaseEventParser``, ``PurchaseCounterByUser``, and ``PurchaseCounterByProduct`` to apply
+different processing on the purchase events and the Workflow ``PurchaseWorkflow`` to connect these MapReduce
+programs in a desired fashion.
+
+The result of the Workflow execution is persisted into a Dataset; the
+application uses the ``createDataset`` method to define the Dataset. We use three datasets:
+``purchaseRecords`` to store the valid parsed purchase events; ``userPurchases`` to store the total purchases
+made by a particular user; and ``productPurchases`` to store the total purchases made for each product.
+The ``Purchase`` class defines the type used to store the parsed purchase events.
+
+The application also adds a custom Workflow action ``EmailNotifier``. When a Workflow executes a custom action,
+it invokes the ``run`` method in the action. In ``EmailNotifier``, we only add a log statement; however it could be
+customized to send emails to the concerned parties.
+
+The ``PurchaseWorkflow`` is scheduled to execute every hour.
+
+Finally, the application adds a service for querying the results from the Dataset.
+
+Let's take a closer look at the Workflow program.
+
+The ``PurchaseWorkflow`` extends an `AbstractWorkflow
+<http://docs.cdap.io/cdap/current/en/reference-manual/javadocs/co/cask/cdap/api/workflow/AbstractWorkflow.html>`__
+class and overrides the ``configure()`` method:
+
+.. code:: java
+
+  public class PurchaseWorkflow extends AbstractWorkflow {
+    @Override
+    protected void configure() {
+      setName("PurchaseWorkflow");
+      setDescription("Workflow to parse the purchase events and count the revenue per user and per product");
+
+      addMapReduce("PurchaseEventParser");
+
+      condition(new PurchaseEventVerifier())
+        .fork()
+          .addMapReduce("PurchaseCounterByUser")
+        .also()
+          .addMapReduce("PurchaseCounterByProduct")
+        .join()
+      .otherwise()
+        .addAction(new NotifyByEmail())
+      .end();
+    }
+  }
+
+In the ``configure()`` method we specify the topology for connecting the programs which will run as a part of
+the Workflow execution. For ``PurchaseWorkflow``, after setting the name and description for the Workflow, we add
```

Comment:
I would not mention name and description here. Rarther: As the first action in the ``PWf``, we add the MapReduce...